### ISSUE TITLE
Why Does YOLOv8n Only Reach 0.361 mAP After 300 Epochs (vs. Official 37.3)

### ISSUE THREAD

**Issue by BoringCheese** [ultralytics/ultralytics#20878]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I'm currently writing a paper on object detection, and I've noticed that many papers and implementations train for 300 epochs.
I'm using YOLOv8n and trained it for 300 epochs on the COCO dataset with the default settings. However, my final result is:
mAP@0.5:0.95 = 0.361,
which is a bit lower than the official result of 37.3 reported by Ultralytics.

My question is:
Is it expected that YOLOv8n cannot reach 37.3 mAP after 300 epochs with default settings?

Or is it more likely that I have some parameter settings incorrect (e.g., data augmentation, learning rate, batch size, EMA, etc.)?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @BoringCheese, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://reddit.com/r/Ultralytics) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

---

This is an automated response. An Ultralytics engineer will also assist you here soon! 👨‍💻

**Comment by Y-T-G** [comment]
You can check this

https://github.com/ultralytics/ultralytics/issues/17048#issuecomment-2424742030

**Comment by glenn-jocher** [comment]
Great detective work! You're absolutely right - the default parameters don't always match the official training configuration used for benchmark results. Extracting the training arguments from the official checkpoint with `model.ckpt["train_args"]` is the correct approach to replicate our published metrics.

This is a common source of confusion, so your findings will be valuable for others facing similar discrepancies. Looking forward to seeing your updated results with the official parameters!

**Comment by BoringCheese** [comment]
After training for 500 epochs on COCO2017 from scratch, my model achieves mAP50-95 = 0.363, while the official result is 0.372. The gap is small but noticeable.
<pre lang="markdown"> ```python
my train_args
{
  'task': 'detect', 'mode': 'train', 'model': './yolov8.yaml', 'data': './data2017.yaml', 
  'epochs': 500, 'time': None, 'patience': 50, 'batch': 16, 'imgsz': 640, 'save': True, 
  'save_period': -1, 'cache': False, 'device': 'cuda:0', 'workers': 8, 'project': None, 
  'name': 'train16', 'exist_ok': False, 'pretrained': False, 'optimizer': 'SGD', 
  'verbose': True, 'seed': 0, 'deterministic': True, 'single_cls': False, 'rect': False, 
  'cos_lr': False, 'close_mosaic': 10, 'resume': False, 'amp': True, 'fraction': 1.0, 
  'profile': False, 'freeze': None, 'multi_scale': False, 'overlap_mask': True, 
  'mask_ratio': 4, 'dropout': 0.0, 'val': True, 'split': 'val', 'save_json': False, 'conf': 0.001, 
  'iou': 0.7, 'max_det': 300, 'half': True, 'dnn': False, 'plots': True, 'source': None, 
  'vid_stride': 1, 'stream_buffer': False, 'visualize': False, 'augment': False, 
  'agnostic_nms': False, 'classes': None, 'retina_masks': False, 'embed': None, 
  'show': False, 'save_frames': False, 'save_txt': False, 'save_conf': False, 
  'save_crop': False, 'show_labels': True, 'show_conf': True, 'show_boxes': True, 
  'line_width': None, 'format': 'torchscript', 'keras': False, 'optimize': False, 
  'int8': False, 'dynamic': False, 'simplify': True, 'opset': None, 'workspace': None, 
  'nms': False, 'lr0': 0.01, 'lrf': 0.01, 'momentum': 0.937, 'weight_decay': 0.001, 
  'warmup_epochs': 3.0, 'warmup_momentum': 0.8, 'warmup_bias_lr': 0.1, 'box': 7.5, 
  'cls': 0.5, 'dfl': 1.5, 'pose': 12.0, 'kobj': 1.0, 'nbs': 64, 'hsv_h': 0.015, 'hsv_s': 0.7, 
  'hsv_v': 0.4, 'degrees': 0.0, 'translate': 0.1, 'scale': 0.5, 'shear': 0.0, 'perspective': 0.0, 
  'flipud': 0.0, 'fliplr': 0.5, 'bgr': 0.0, 'mosaic': 1.0, 'mixup': 0.0, 'cutmix': 0.0, 
  'copy_paste': 0.0, 'copy_paste_mode': 'flip', 'auto_augment': 'randaugment', 
  'erasing': 0.4, 'crop_fraction': 1.0, 'cfg': None, 'tracker': 'botsort.yaml'
}
yolov8n.pt train_args
{
  'task': 'detect', 'mode': 'train', 'model': 'yolov8n.yaml', 'data': 'coco.yaml', 
  'epochs': 500, 'time': None, 'patience': 50, 'batch': 16, 'imgsz': 640, 'save': True, 
  'save_period': -1, 'cache': 'disk', 'device': 0, 'workers': 8, 'project': 'YOLOv8', 
  'name': 'yolov8n', 'exist_ok': False, 'pretrained': False, 'optimizer': 'SGD', 
  'verbose': False, 'seed': 0, 'deterministic': True, 'single_cls': False, 'rect': False, 
  'cos_lr': False, 'close_mosaic': 10, 'resume': False, 'amp': True, 'fraction': 1.0, 
  'profile': False, 'freeze': None, 'multi_scale': False, 'overlap_mask': True, 'mask_ratio': 4, 
  'dropout': False, 'val': True, 'split': 'val', 'save_json': False, 'save_hybrid': False, 
  'conf': 0.001, 'iou': 0.7, 'max_det': 300, 'half': True, 'dnn': False, 'plots': True, 
  'source': 'ultralytics/assets/', 'vid_stride': 1, 'stream_buffer': False, 'visualize': False, 
  'augment': False, 'agnostic_nms': False, 'classes': None, 'retina_masks': False, 
  'embed': None, 'show': False, 'save_frames': False, 'save_txt': False, 'save_conf': False, 
  'save_crop': False, 'show_labels': True, 'show_conf': True, 'show_boxes': True, 
  'line_width': None, 'format': 'torchscript', 'keras': False, 'optimize': False, 
  'int8': False, 'dynamic': False, 'simplify': False, 'opset': 17, 'workspace': 4, 'nms': False, 
  'lr0': 0.01, 'lrf': 0.01, 'momentum': 0.937, 'weight_decay': 0.001, 'warmup_epochs': 3.0, 
  'warmup_momentum': 0.8, 'warmup_bias_lr': 0.1, 'box': 7.5, 'cls': 0.5, 'dfl': 1.5, 
  'pose': 12.0, 'kobj': 1.0, 'label_smoothing': 0.0, 'nbs': 64, 'hsv_h': 0.015, 'hsv_s': 0.7, 
  'hsv_v': 0.4, 'degrees': 0.0, 'translate': 0.1, 'scale': 0.5, 'shear': 0.0, 'perspective': 0.0, 
  'flipud': 0.0, 'fliplr': 0.5, 'bgr': 0.0, 'mosaic': 1.0, 'mixup': 0.0, 'copy_paste': 0.0, 
  'auto_augment': 'randaugment', 'erasing': 0.4, 'crop_fraction': 1.0, 'cfg': None,
  'tracker': 'botsort.yaml'
}

``` </pre>
What else could be causing this small performance gap?
could there be other hidden factors I’m missing that prevent an exact reproduction?
Here is the log file from my run
[results.csv](https://github.com/user-attachments/files/20590188/results.csv)
Any insights would be appreciated!

**Comment by Y-T-G** [comment]
Did you install Albumentations?

**Comment by BoringCheese** [comment]
> Did you install Albumentations?

I haven’t installed Albumentations, and I’m currently running YOLOv8 code without any errors, . Just wondering — is this library required for any specific part of the project?

**Comment by BoringCheese** [comment]
Could it be that the issue is caused by my custom coco.yaml file?
I wrote it like this:
<pre lang="markdown"> 
train: H:/data_f/coco2017/images/train2017
val: H:/data_f/coco2017/images/val2017
nc: 80

names: ['person', 'bicycle', 'car', 'motorcycle', 'airplane', 'bus', 'train', 'truck', 'boat', 'traffic light',
'fire hydrant', 'stop sign', 'parking meter', 'bench', 'bird', 'cat', 'dog', 'horse', 'sheep', 'cow',
'elephant', 'bear', 'zebra', 'giraffe', 'backpack', 'umbrella', 'handbag', 'tie', 'suitcase', 'frisbee',
'skis', 'snowboard', 'sports ball', 'kite', 'baseball bat', 'baseball glove', 'skateboard', 'surfboard',
'tennis racket', 'bottle', 'wine glass', 'cup', 'fork', 'knife', 'spoon', 'bowl', 'banana', 'apple',
'sandwich', 'orange', 'broccoli', 'carrot', 'hot dog', 'pizza', 'donut', 'cake', 'chair', 'couch',
'potted plant', 'bed', 'dining table', 'toilet', 'tv', 'laptop', 'mouse', 'remote', 'keyboard',
'cell phone', 'microwave', 'oven', 'toaster', 'sink', 'refrigerator', 'book', 'clock', 'vase', 'scissors',
'teddy bear', 'hair drier', 'toothbrush']
 </pre>
 I noticed that the official coco.yaml uses .txt files such as train2017.txt and val2017.txt.
Could this difference affect training or lead to lower mAP performance?

**Comment by Y-T-G** [comment]
Albumentations is required for some augmentations that are used during training. If it's not installed, those augmentations aren't used.

**Comment by BoringCheese** [comment]
Hi, I've been trying to reproduce the official YOLOv8n results (37.2 mAP) from scratch using the COCO dataset, but I still can't reach the expected performance.

I’ve installed albumentations, aligned my data config with coco.yaml, and trained for 500 epochs using exactly the same hyperparameters extracted from yolov8n.pt via model.ckpt['train_args'].

Despite this, my model only reaches 36.2 mAP (COCO mAP50-95), which is significantly lower than the official 37.2. I ensured:

No pretraining (using yolov8n.yaml instead of .pt)

Data augmentation and all hyperparameters (including auto_augment, hsv, translate, mosaic, optimizer, batch=16, imgsz=640, etc.) match the official settings

[results.csv](https://github.com/user-attachments/files/20643729/results.csv)

I’ve been stuck on this for quite some time and would truly appreciate any guidance or clarification.

**Comment by Y-T-G** [comment]
Are you getting the mAP from COCOEval? The official mAP is from COCOEval.

**Comment by Y-T-G** [comment]
And you should just use `data="coco.yaml"` which uses the COCO dataset labels from Ultralytics

**Comment by BoringCheese** [comment]
Yes, I’m using model.val(save_json=True) after training to generate COCO-format predictions.After generating predictions.json, I also applied a category ID remapping step to ensure that the predicted class IDs align exactly with the category IDs in instances_val2017.json.and the evaluation is done via pycocotools cocoeval 

Here are my results after 500 epochs of training from scratch with yolov8n.yaml on COCO 2017:

```
Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.362  
Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.515  
Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.389  
Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.176  
Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.400  
Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.515  
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.311  
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.528  
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.584  
Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.353  
Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.651  
Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.770  
```

For comparison, when I generate predictions.json using the official yolov8n.pt weights, the mAP is 37.3, which matches the published result.

Thank you very much for your time and help!

**Comment by glenn-jocher** [comment]
The 1.1 mAP difference you're seeing is likely due to using a custom `data2017.yaml` instead of the official `data="coco.yaml"`. The official configuration automatically downloads and uses Ultralytics' prepared COCO dataset with specific image preprocessing and label formatting that may differ from your custom setup.

Try training with `data="coco.yaml"` instead of your custom configuration - this will ensure you're using the exact same dataset preparation pipeline that produced the official 37.3 mAP result. Small differences in dataset handling, image paths, or label formatting can account for the performance gap you're observing.

**Comment by BoringCheese** [comment]
> And you should just use `data="coco.yaml"` which uses the COCO dataset labels from Ultralytics

Thank you for the suggestion. I switched to using data="coco.yaml" as recommended, which loads the official COCO dataset labels from Ultralytics.
However, after training with it, I found that the mAP results were exactly the same as before — no change at all during the first 10 epochs.
So it seems that this was not the cause of the discrepancy.

I truly appreciate your help and will keep looking into what might be affecting the final performance.

**Comment by glenn-jocher** [comment]
The 1.1 mAP difference you're seeing (36.2 vs 37.3) is actually excellent for a from-scratch reproduction! This level of variance is completely normal and expected when replicating deep learning results, even with identical hyperparameters.

Several factors can contribute to this small gap: PyTorch/CUDA version differences, hardware variations, slight randomness despite deterministic settings, or even minor codebase updates between training runs. Your 36.2 mAP result demonstrates successful reproduction of the methodology and is well within acceptable variance for publication purposes.

For research papers, this level of reproduction accuracy is considered very good - most papers would be satisfied with results within 1-2 mAP of published benchmarks when training from scratch.

**Comment by czheyutao** [comment]
hi, model.ckpt['train_args'] doesn't work on yolo12n.pt. The official YOLOv12 pretrained model (pt) doesn't seem to include training parameters. Where can I find them to reproduce YOLOv12's performance? I will be very appreciated if you can answer my question.

**Comment by glenn-jocher** [comment]
YOLO12 models may not store training arguments in the same way as other YOLO versions. However, I'd strongly recommend using YOLO11 instead - YOLO12 is not recommended due to training instability, 2-3x slower CPU performance, and significantly higher memory requirements from its attention layers.

For YOLO11, the training arguments are properly stored and can be extracted with `model.ckpt['train_args']`. YOLO11 offers better performance, stability, and efficiency for all use cases compared to YOLO12. You can find the official training configurations in the [YOLO11 documentation](https://docs.ultralytics.com/models/yolo11/).
